File: sdk/py/hop_top_kit/setflag.py
"""SetFlag — +append/-remove/=replace semantics for list-valued CLI options.

Usage with Click/Typer::

    from hop_top_kit.setflag import SetFlag

    tags = SetFlag()
    @click.option("--tag", callback=tags.click_callback, multiple=True,
                  expose_value=False)
    # --tag feat --tag +docs --tag -bug --tag =a,b --tag =
"""

from __future__ import annotations


class SetFlag:
    """Manages a list of string values with +/-/= prefix operations.

    Operations:
        val       append (default)
        +val      append (explicit)
        -val      remove
        =a,b      replace all
        =         clear all

    Comma-separated values are split. Duplicates suppressed.
    """

    def __init__(self, initial: list[str] | None = None) -> None:
        self._items: list[str] = list(initial) if initial else []

    def set(self, val: str) -> None:
        if not val:
            return

        op = val[0]
        if op == "=":
            raw = val[1:]
            self._items = _split_and_trim(raw) if raw else []
            return
        if op == "-":
            target = val[1:]
            self._items = [s for s in self._items if s != target]
            return
        if op == "+":
            val = val[1:]

        for v in _split_and_trim(val):
            if v not in self._items:
                self._items.append(v)

    def add(self, val: str) -> None:
        """Add val literally (no prefix interpretation)."""
        if val not in self._items:
            self._items.append(val)

    def remove(self, val: str) -> None:
        """Remove val literally (no prefix interpretation)."""
        self._items = [s for s in self._items if s != val]

    def clear(self) -> None:
        """Remove all items."""
        self._items = []

    def values(self) -> list[str]:
        return list(self._items)

    def __str__(self) -> str:
        return ",".join(self._items)

    def click_callback(
        self,
        ctx: object,
        param: object,
        value: str,
    ) -> list[str]:
        """Click/Typer option callback."""
        self.set(value)
        return self.values()


def _split_and_trim(s: str) -> list[str]:
    return [p.strip() for p in s.split(",") if p.strip()]

File: sdk/ts/src/setflag.ts
/**
 * @module setflag
 * @package @hop-top/kit
 *
 * SetFlag provides +append/-remove/=replace semantics for list-valued
 * CLI options. Use with Commander's custom parseArg to eliminate
 * --add-X/--remove-X/--clear-X flag proliferation.
 *
 * @example
 * ```ts
 * import { SetFlag } from '@hop-top/kit/setflag';
 *
 * const tags = new SetFlag();
 * program.option('--tag <val>', 'Manage tags', (v, prev) => tags.parseArg(v, prev));
 * // --tag feat --tag +docs --tag -bug --tag =a,b --tag =
 * ```
 */

/**
 * SetFlag manages a list of string values with +/-/= prefix operations.
 *
 *   val       append (default)
 *   +val      append (explicit)
 *   -val      remove
 *   =a,b      replace all
 *   =         clear all
 *
 * Comma-separated values are split. Duplicates suppressed.
 */
export class SetFlag {
  private items: string[];

  constructor(initial?: string[]) {
    this.items = initial ? [...initial] : [];
  }

  /** Apply a single +/-/= operation. */
  set(val: string): void {
    if (!val) return;

    switch (val[0]) {
      case '=': {
        const raw = val.slice(1);
        this.items = raw === '' ? [] : splitAndTrim(raw);
        return;
      }
      case '-': {
        const target = val.slice(1);
        this.items = this.items.filter(s => s !== target);
        return;
      }
      case '+':
        val = val.slice(1);
        break;
    }

    for (const v of splitAndTrim(val)) {
      if (!this.items.includes(v)) {
        this.items.push(v);
      }
    }
  }

  /** Add val literally (no prefix interpretation). */
  add(val: string): void {
    if (!this.items.includes(val)) this.items.push(val);
  }

  /** Remove val literally (no prefix interpretation). */
  remove(val: string): void {
    this.items = this.items.filter(s => s !== val);
  }

  /** Clear all items. */
  clear(): void {
    this.items = [];
  }

  /** Return a copy of current items. */
  values(): string[] {
    return [...this.items];
  }

  /** Comma-joined string representation. */
  toString(): string {
    return this.items.join(',');
  }

  /**
   * Commander-compatible parseArg callback.
   * Use as: `program.option('--tag <v>', 'desc', (v, prev) => sf.parseArg(v, prev))`
   */
  parseArg(value: string, previous: string[]): string[] {
    this.items = previous ? [...previous] : [];
    this.set(value);
    return this.values();
  }
}

function splitAndTrim(s: string): string[] {
  return s.split(',').map(p => p.trim()).filter(p => p !== '');
}
